scripts/validate_docstrings.py

```diff
@@ -490,10 +491,41 @@ def yields(self):
     @property
     def method_source(self):
         try:
-            return inspect.getsource(self.obj)
+            source = inspect.getsource(self.obj)
+            # Remove common indentation.
+            first_spaces = re.match(' +', source)
+            if first_spaces:
+                first_spaces = first_spaces.group(0)
+                source = re.sub('^' + first_spaces, '', source)
+                source = re.sub(r'\n' + first_spaces, r'\n', source)
+            return source
         except TypeError:
             return ''
 
+    @property
+    def method_returns(self):
+        tree = ast.parse(self.method_source).body
+        if tree:
+            root = tree[0]
+
+            def gather_returns(node):
```

Comment:
Hmm is there any way to simplify what is going on here? Is there no way to extract the `.returns` from the AST directly instead of having to recurse everything?

Reply:
In our case, there isn't.

The AST provides the `.walk()` method to iterate through nodes. According to [the doc](https://docs.python.org/3/library/ast.html#ast.walk), it
```
is useful if you only want to modify nodes in place and don’t care about the context.
```
Without context, there is no way to tell if a return node is inside a nested inner function or not, which is needed for one of the [requirements](https://github.com/pandas-dev/pandas/pull/25008#issuecomment-460679572) raised by @datapythonista. 

I explain this a bit further [here](https://github.com/pandas-dev/pandas/pull/25008#discussion_r254978169).